### ISSUE TITLE
💡 Feature:  checkmk 2.3 support

### ISSUE THREAD

**Issue by tokajer** [WagnerAG/checkmk_fortigate#1]
### Feature Description

Hi,

its not working with 2.3 versions, a rewrite for 2.3 would be nice.

also some gui error when i activate the plugin.

Agent based plug-ins (deprecated): 28, Agents: 1, Checks' man pages (deprecated): 20, Legacy check plug-ins (deprecated): 1, Legacy GUI extensions (deprecated): 14

keep up the good work!

**Comment by crpb** [comment]
My gui still works on `Checkmk Raw Edition 2.3.0p19` but the cli doesn't like it

```
OMD[sitename]:~$ cmk --debug --check-discovery $myhost
/omd/sites/sitename/lib/python3.12/site-packages/pydantic/_internal/_fields.py:149: UserWarning: Field "model_name" has conflict with protected namespace "model_".

You may be able to resolve this warning by setting `model_config['protected_namespaces'] = ()`.
  warnings.warn(
/omd/sites/sitename/lib/python3.12/site-packages/pydantic/_internal/_fields.py:149: UserWarning: Field "model_number" has conflict with protected namespace "model_".

You may be able to resolve this warning by setting `model_config['protected_namespaces'] = ()`.
  warnings.warn(
Traceback (most recent call last):
  File "/omd/sites/sitename/bin/cmk", line 101, in <module>
    errors = config.load_all_plugins(
             ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/omd/sites/sitename/lib/python3/cmk/base/config.py", line 1420, in load_all_plugins
    errors = agent_based_register.load_all_plugins()
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/omd/sites/sitename/lib/python3/cmk/base/api/agent_based/register/_discover.py", line 53, in load_all_plugins
    raise exception
  File "/omd/sites/sitename/lib/python3/cmk/utils/plugin_loader.py", line 51, in load_plugins
    yield name, importlib.import_module(full_name)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/omd/sites/sitename/lib/python3.12/importlib/__init__.py", line 90, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap>", line 1387, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1360, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1331, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 935, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 995, in exec_module
  File "<frozen importlib._bootstrap>", line 488, in _call_with_frames_removed
  File "/omd/sites/sitename/local/lib/python3/cmk/base/plugins/agent_based/fortios_interface.py", line 106, in <module>
    class VdomDataList(BaseModel):
  File "/omd/sites/sitename/lib/python3.12/site-packages/pydantic/_internal/_model_construction.py", line 92, in __new__
    private_attributes = inspect_namespace(
                         ^^^^^^^^^^^^^^^^^^
  File "/omd/sites/sitename/lib/python3.12/site-packages/pydantic/_internal/_model_construction.py", line 316, in inspect_namespace
    raise TypeError("To define root models, use `pydantic.RootModel` rather than a field called '__root__'")
TypeError: To define root models, use `pydantic.RootModel` rather than a field called '__root__'
OMD[sitename]:~$
```

Didn't

**Comment by 7meis** [comment]
Hello, until we have a version that is Checkmk 2.3.0 compatible, I have set the maximum version to 2.3.0b1.

**Comment by crpb** [comment]
> This issue has been automatically marked as stale because it has not had recent activity. It will be closed if no further activity occurs.

SHOO!!

**Comment by pandalec** [comment]
Any progress on this? I am using checkmk 2.3.0pXX and I'd love to test this plugin

**Comment by 7meis** [comment]
Hello guys, the 2.3.0 mkp is available.
Can you please try the MKP for Version 2.3.0. See in branch "cmk23"
Direct download link: https://github.com/WagnerAG/checkmk_fortigate/blob/cmk23/checkmk_fortios-2.3.0.mkp 
Any feedback is welcome.

**Comment by pandalec** [comment]
Thank you! First thing I found, the version inside the mkp (info/info.json) differs from the download filename (still 1.0.0). My automation doesn't like this ;)

**Comment by 7meis** [comment]
Please find here https://github.com/WagnerAG/checkmk_fortigate/blob/cmk23/checkmk_fortios-2.3.0_v1.0.1.mkp version 1.0.1 as mkp. ;-)

**Comment by INIT-ALNY** [comment]
Hello!
In advance, i'm kinda new to CheckMK and plugins. 
Running CheckMK Raw edition 2.3.0p28 
I get errors when running a discovery on a FortiGate
Not sure if it's my setup, or if it's the plugin that causes these errors

Starting job...
WARNING: Parsing of section fortios_ipsec failed - please submit a crash report! (Crash-ID: 02ca19e4-1aa1-11f0-af5d-5b88b6a3f9ed)
WARNING: Parsing of section fortios_interfaces_cmdb failed - please submit a crash report! (Crash-ID: 02ca19e5-1aa1-11f0-af5d-5b88b6a3f9ed)
WARNING: Parsing of section fortios_dhcp_lease failed - please submit a crash report! (Crash-ID: 02ca19e6-1aa1-11f0-af5d-5b88b6a3f9ed)
WARNING: Parsing of section fortios_managed_ap failed - please submit a crash report! (Crash-ID: 02ca19e7-1aa1-11f0-af5d-5b88b6a3f9ed)
  WARNING: Exception in discovery function of check plug-in 'fortios_interfaces': 'NoneType' object has no attribute 'get'
Completed.

I still get information from the API, and the sensors/services works fine. But not sure what is causing this 
Is this something anyone has encountered? 😊

Let me know, if other information is required, or if this is even the right place to ask 😅
Thanks for the great work!

**Comment by dampfhamm3r** [comment]
> Hello! In advance, i'm kinda new to CheckMK and plugins. Running CheckMK Raw edition 2.3.0p28 I get errors when running a discovery on a FortiGate Not sure if it's my setup, or if it's the plugin that causes these errors
> 
> Starting job... WARNING: Parsing of section fortios_ipsec failed - please submit a crash report! (Crash-ID: 02ca19e4-1aa1-11f0-af5d-5b88b6a3f9ed) WARNING: Parsing of section fortios_interfaces_cmdb failed - please submit a crash report! (Crash-ID: 02ca19e5-1aa1-11f0-af5d-5b88b6a3f9ed) WARNING: Parsing of section fortios_dhcp_lease failed - please submit a crash report! (Crash-ID: 02ca19e6-1aa1-11f0-af5d-5b88b6a3f9ed) WARNING: Parsing of section fortios_managed_ap failed - please submit a crash report! (Crash-ID: 02ca19e7-1aa1-11f0-af5d-5b88b6a3f9ed) WARNING: Exception in discovery function of check plug-in 'fortios_interfaces': 'NoneType' object has no attribute 'get' Completed.
> 
> I still get information from the API, and the sensors/services works fine. But not sure what is causing this Is this something anyone has encountered? 😊
> 
> Let me know, if other information is required, or if this is even the right place to ask 😅 Thanks for the great work!

Hi
Difficult to say from the output.
I assume you followed the steps mentioned in the [README.md](https://github.com/WagnerAG/checkmk_fortigate/blob/cmk23/README.md) for the special agent setup?

Can you post the output of the following command:
` /opt/omd/sites/cmk/bin/python3 agent_fortios.py --api-token <TOKEN> --port 8443 --no-cert-check <HOST_IP>`

Also, an output of a crash report could be helpful.

What type of Fortigate do you have? Physical or VM?

**Comment by INIT-ALNY** [comment]
Hi dampfhamm3r,
Thanks for the fast response 😁

To answer your questions first.
I did follow the Readme, and created the API token on the Fortigates, as specified, and also setup the rule in CheckMK to get communication up and running. Atleast I assume i did this all correct, otherwise I would get any information from the Fortigate at all right? Correct me if i'm wrong 😊
As to what type of Fortigates we have. All of them but one is physical Fortigates. We have a single Fortigate appliance in the Cloud. Although more or less, all of the Fortigates produce the same crash reports, both physical as appliance. There are some with more crash reports, and others with less. Not sure what causes the difference.
To note, we have a wide range of different models (40F, 60F, 100F and so on), not sure if this could explain some of the differences of what we see. They all to my knowledge run v7.2.10.

In regards to the following command: /opt/omd/sites/cmk/bin/python3 agent_fortios.py --api-token <TOKEN> --port 8443 --no-cert-check <HOST_IP>
The agent_fortios.py doesn't seem to be present in this location. I assumed that "cmk" was the name of the site, so i tried with both:
- /opt/omd/sites/cmk/bin/python3
- /opt/omd/sites/<MYSITE>/bin/python3
But non of them had the script here.
I tried to search for that specific file and found it to be located in: /opt/omd/sites/<MYSITE>/local/share/check_mk/web/plugins/wato

Not sure if this has to with me manually importing the plugin with MKP, since i don'h have the enterprise version that includes the DCD feature. But nonetheless the output of the following command:  /opt/omd/sites/<MYSITE>/local/share/check_mk/web/plugins/wato agent_fortios.py --api-token <TOKEN> --port 8443 --no-cert-check <HOST_IP>

Gave me this:

Traceback (most recent call last):
  File "/opt/omd/sites/<MYSITE>/local/share/check_mk/web/plugins/wato/agent_fortios.py", line 25, in <module>
    from cmk.gui.i18n import _
ModuleNotFoundError: No module named 'cmk'

In regards to a crash report, i have taken one as an example. There are many of them 😅

{
    "core": "nagios",
    "crash_type": "check",
    "details": {
        "check_output": "check failed - please submit a crash report!",
        "check_type": "discover",
        "description": "Check_MK Discovery",
        "enforced_service": false,
        "host": "<MYHOST>",
        "inline_snmp": false,
        "is_cluster": false
    },
    "edition": "cre",
    "exc_traceback": [
        [
            "/omd/sites/<MYSITE>/lib/python3/cmk/base/modes/check_mk.py",
            1663,
            "mode_check_discovery",
            "checks_result = execute_check_discovery("
        ],
        [
            "/omd/sites/<MYSITE>/lib/python3/cmk/checkengine/discovery/_impl.py",
            147,
            "execute_check_discovery",
            "services = get_host_services("
        ],
        [
            "/omd/sites/<MYSITE>/lib/python3/cmk/checkengine/discovery/_autodiscovery.py",
            609,
            "get_host_services",
            "**_get_node_services("
        ],
        [
            "/omd/sites/<MYSITE>/lib/python3/cmk/checkengine/discovery/_autodiscovery.py",
            663,
            "_get_node_services",
            "discovered_services = discover_services("
        ],
        [
            "/omd/sites/<MYSITE>/lib/python3/cmk/checkengine/discovery/_services.py",
            125,
            "discover_services",
            "{"
        ],
        [
            "/omd/sites/<MYSITE>/lib/python3/cmk/checkengine/discovery/_services.py",
            186,
            "_discover_plugins_services",
            "yield from plugin.function(check_plugin_name, **kwargs)"
        ],
        [
            "/omd/sites/<MYSITE>/lib/python3/cmk/base/checkers.py",
            905,
            "__discovery_function",
            "yield from ("
        ],
        [
            "/omd/sites/<MYSITE>/lib/python3/cmk/base/checkers.py",
            905,
            "<genexpr>",
            "yield from ("
        ],
        [
            "/omd/sites/<MYSITE>/lib/python3/cmk/base/api/agent_based/register/check_plugins.py",
            75,
            "filtered_generator",
            "for element in generator(*args, **kwargs):"
        ],
        [
            "/omd/sites/<MYSITE>/local/lib/python3/cmk/base/plugins/agent_based/fortios_interface.py",
            152,
            "discovery_fortios_interfaces",
            "interface_cmdb = section_fortios_interfaces_cmdb.get(interface.name)"
        ]
    ],
    "exc_type": "AttributeError",
    "exc_value": "'NoneType' object has no attribute 'get'",
    "id": "40547c61-1bfd-11f0-af60-5b88b6a3f9ed",
    "local_vars": "REDACTED - Was a bunch of random numbers and letters",
    "os": "Ubuntu 22.04.5 LTS",
    "python_paths": [
        "/opt/omd/versions/2.3.0p28.cre/bin",
        "/omd/sites/<MYSITE>/local/lib/python3",
        "/omd/sites/<MYSITE>/lib/python3/cloud",
        "/omd/sites/<MYSITE>/lib/python312.zip",
        "/omd/sites/<MYSITE>/lib/python3.12",
        "/omd/sites/<MYSITE>/lib/python3.12/lib-dynload",
        "/omd/sites/<MYSITE>/lib/python3.12/site-packages",
        "/omd/sites/<MYSITE>/lib/python3"
    ],
    "python_version": "3.12.3 (main, Feb 26 2025, 12:27:18) [GCC 13.2.0]",
    "time": 1744943497.0444481,
    "version": "2.3.0p28"
}

Hope this information helps. And again i'm sorry, my knowledge about CheckMK plugins are basically zero 😅
I appreciate the help so far 

Let me know if you need any other information 😊

**Comment by tokajer** [comment]
@INIT-ALNY  please update to 7.2.11 asap if not already running. I would advise to check the version.
If you need assistance you can write me ;)

@dampfhamm3r i will check it next week if I have the same error.

**Comment by 7meis** [comment]
Hello @INIT-ALNY

Can you please try with this path:
 /opt/omd/sites/<SITE>/local/share/check_mk/agents/special/agent_fortios --api-token <TOKEN> --port <PORT> --no-cert-check <HOST_IP>

The path you tried was the WATO (Web Plugin), this obviuously will not work: opt/omd/sites//local/share/check_mk/web/plugins/wato agent_fortios.py

Please send the output of the command to: opensource@wagner.ch so you don't have to post it here.

**Comment by tokajer** [comment]
hi i have the same error. with 2.3p6
Starting job...
WARNING: Parsing of section fortios_dhcp_scope failed - please submit a crash report! (Crash-ID: db2f2330-201c-11f0-8b64-f5f7f48d6c5d)
WARNING: Parsing of section fortios_ipsec failed - please submit a crash report! (Crash-ID: db2f2331-201c-11f0-8b64-f5f7f48d6c5d)
WARNING: Parsing of section fortios_managed_ap failed - please submit a crash report! (Crash-ID: db2f2332-201c-11f0-8b64-f5f7f48d6c5d)
WARNING: Parsing of section fortios_dhcp_lease failed - please submit a crash report! (Crash-ID: db2f2333-201c-11f0-8b64-f5f7f48d6c5d)
Completed.

**Comment by standadHD** [comment]
Traceback (most recent call last):
  File "/opt/omd/sites/standiCC/local/share/check_mk/agents/special/agent_fortios", line 32, in <module>
    from cmk.special_agents.v0_unstable.agent_common  import (
ModuleNotFoundError: No module named 'cmk'

**Comment by 7meis** [comment]
This is, when you execute the command as root. Please use the site user ->
su - <site> and then execute the command.

**Comment by standadHD** [comment]
<<<fortios_dhcp_scope:sep(0)>>>
dhcp config is read out correctly, without errors

<<<fortios_dhcp_lease:sep(0)>>>
{"action": "", "build": 1706, "http_method": "GET", "name": "dhcp", "path": "system", "results": [{"expire_time": 1745565254, "....

leases read out correctly

"success", "vdom": "root", "version": "v7.2.10"}
ERROR 2025-04-25 06:23:39 agent_fortios: Collecting section: sensors failed. Reason: HTTP status not 200; error: (404) Not Found
HTTP status error: Spec name: sensors failed. Reason: HTTP status not 200; error: (404) Not Found
ERROR 2025-04-25 06:23:39 agent_fortios: Collecting sensors failed: monitor/system/sensor-info


(otmitted the detailed output due security reasons)

**Comment by 7meis** [comment]
The 404 error is not a big deal. Not all models support this endpoint.
Can you please redact the json output of the check who is crashing and send
it to me?

**Comment by mhofmann-hb** [comment]
We are on FortiOs 7.4.7 and it seems that sensor-info is not available anymore:

{
      "path":"system",
      "name":"sensor-info",
      "action":"select",
      "supported":false,
      "access_group":"sysgrp.cfg",
      "summary":"Retrieve system sensor status.",
      "request":{
        "http_method":"GET"
      },

I ithink that maybe "supported":false is the problem...

curl --insecure -H "Accept: application/json" -H "Authorization: Bearer bla" https://1.2.3.4:8443/api/v2/monitor/system/sensor-info
{
  "path":"system",
  "name":"sensor-info",
  "action":"",
  "serial":"FGVM4VTM20000567",
  "version":"v7.4.7",
  "build":2731,
  "status":"error",
  "http_status":404
